### ISSUE TITLE
Explaination for CanadianComputingCompetition/2018/Senior/balancedTrees.py?

### ISSUE THREAD

**Issue by KokeCacao** [zecookiez/CanadianComputingCompetition#2]
Hi Zeyu,

As I am suffering from CCC 2020 preparation, thank you for creating this repo. I am new to IO and my interests are mainly on Deep Learning, therefore I cannot understand complex math algorithms like `balancedTrees.py`. Can you explain the solution to balancedTrees more? Did you create `balancedTrees.py` on your own? What was the inspiration? How did you come up with the series?

Best,
Hanke Chen
12th grader

**Comment by zecookiez** [comment]
Hi Hanke,

Yes, I wrote this solution by myself. My initial idea was to map the problem out into a recursive sequence, where the `i-th` term returned the number of balanced trees with weight `i`. This part is mostly done due to the wording of the problem statement itself, so coming up with it wasn't as difficult. The recursion passed the 5/15 point subtask.

For the second and third subtask, the trick is to apply dynamic programming. If we simulated the recursion on paper, it became obvious that many terms were recalculated multiple times. By applying memoization this passed all the additional subtasks.

The final observation is the number of identical terms when doing floor division. If there is a way to calculate the number of repeated terms, we could easily multiply that with the output from the terms. And turns out there is a formula! The formula is described as `floor(n / i) - floor(n / (i + 1))`. This gets all the remaining points.

Hopefully this clears some of your questions up :)

Cheers,
Zeyu Chen
